tiledb/sm/filter/webp_filter.cc

```diff
@@ -0,0 +1,324 @@
+      pixel_depth = format_ < WebpInputFormat::WEBP_RGBA ? 3 : 4;
+  // X should be divisible by colorspace pixel_depth or RGB values will skew.
+  if (extent_x % pixel_depth != 0) {
+    throw StatusException(Status_FilterError(
+        pixel_depth == 3 ?
+            "Colorspace with no alpha must use extents divisible by 3" :
+            "Colorspace with alpha must use extents divisible by 4"));
+  }
+
+  for (const auto& i : input_parts) {
+    auto data = static_cast<const uint8_t*>(i.data());
+    // Number of bytes encoded; Encoded result data buffer.
+    size_t enc_size = 0;
+    uint8_t* result = nullptr;
+    // Cleanup allocated data when we leave scope.
+    ScopedExecutor cleanup([&]() {
+      if (enc_size > 0 && result != nullptr)
+        WebPFree(result);
+    });
+    // We divide extent_x by colorspace value count to get pixel-width of image,
+    // extent_x currently represents row stride.
+    switch (format_) {
+      case WebpInputFormat::WEBP_RGB:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessRGB(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeRGB(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_RGBA:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessRGBA(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeRGBA(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_BGR:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessBGR(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeBGR(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_BGRA:
+        if (lossless_) {
+          enc_size = WebPEncodeLosslessBGRA(
+              data, extent_x / pixel_depth, extent_y, extent_x, &result);
+        } else {
+          enc_size = WebPEncodeBGRA(
+              data,
+              extent_x / pixel_depth,
+              extent_y,
+              extent_x,
+              quality_,
+              &result);
+        }
+        break;
+      case WebpInputFormat::WEBP_NONE:
+        throw StatusException(Status_FilterError(
+            "Filter option TILEDB_FILTER_WEBP_FORMAT must be set"));
+    }
+
+    // Check if encoding failed.
+    if (enc_size == 0) {
+      throw StatusException(Status_FilterError("Error encoding image data"));
```

Comment:
nit: `Status_` objects can now be thrown directly

Reply:
Thanks for the review! Initially when I threw `Status_FilterError` directly I had this output for any / all exceptions thrown
```
due to unexpected exception with message:
  C API: TileDB Internal: unknown exception type; no further information
  ```
  
So I made a small change to [exception_wrapper.h](https://github.com/TileDB-Inc/TileDB/commit/609b0cf07123cef4a321cfe0b1228e2f794e06d1#diff-c567adabd29ac47cccbe96a7a728bd19e34ac5617f00f5c077e4f34e28983cbf) in my last commit and now we have correct output for exceptions but if we would rather do something like [throw_config_exception](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/config/config.cc#L52) I can revert my changes to `exception_wrapper.h` and do this instead.

I also wanted to point out a change I made to [dimension.h](https://github.com/TileDB-Inc/TileDB/commit/609b0cf07123cef4a321cfe0b1228e2f794e06d1#diff-deed9af6b2d0a5be36a7435e4379a5eb3f1371b7d338262d91796b3190b8d2adR321) which fixed failing test cases for dimensions of type `int8_t` and `uint8_t`. There seemed to be overflow happening between the multiplication in this function and casting back down to the return type `T` which results in a segfault during write, but I can revert this and take another look if needed. I wrote a small (~20 line) test case that reproduced the segfault on a dense array that was not using the WebP filter.

Reply:
I reverted changes around exceptions after our sync and reproduced / fixed failing CI job locally using manylinux docker container. :+1: 

> I also wanted to point out a change I made to [dimension.h](https://github.com/TileDB-Inc/TileDB/commit/609b0cf07123cef4a321cfe0b1228e2f794e06d1#diff-deed9af6b2d0a5be36a7435e4379a5eb3f1371b7d338262d91796b3190b8d2adR321) which fixed failing test cases for dimensions of type `int8_t`. There seemed to be overflow happening between the multiplication in this function and casting back down to the return type `T` which results in a segfault during write, but I can revert this and take another look if needed. I wrote a small (~20 line) [test case](https://gist.github.com/shaunrd0/affca045ec3992d5062ca8235c35060b) that reproduced the segfault on a dense array that was not using the WebP filter.

@KiterLuc just checking if you have any input on this? It looks like the change I made in `dimension.h` involves code introduced by #2303.